src/daemon/distribution.c

```diff
@@ -0,0 +1,291 @@
+                                        double factor) {
+  bucket_t *buckets = (bucket_t *)calloc(num_buckets, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  double multiplier = initial_size;
+
+  for (size_t i = 0; i < num_buckets - 1; ++i) {
+    buckets[i].max_boundary = factor * multiplier;
+    multiplier *= initial_size;
+  }
+
+  buckets[num_buckets - 1].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+static bucket_t *bucket_new_custom(size_t num_boundaries,
+                                   const double *custom_buckets_boundaries) {
+  bucket_t *buckets = (bucket_t *)calloc(num_boundaries + 1, sizeof(bucket_t));
+
+  if (buckets == NULL) {
+    return NULL;
+  }
+
+  if (custom_buckets_boundaries[0] <= 0) {
+    free(buckets);
+    errno = EINVAL;
+    return NULL;
+  }
+
+  buckets[0].max_boundary = custom_buckets_boundaries[0];
+
+  for (size_t i = 1; i < num_boundaries; ++i) {
+    if (custom_buckets_boundaries[i] <= 0 ||
+        custom_buckets_boundaries[i - 1] >= custom_buckets_boundaries[i]) {
+      free(buckets);
+      errno = EINVAL;
+      return NULL;
+    }
+
+    buckets[i].max_boundary = custom_buckets_boundaries[i];
+  }
+
+  buckets[num_boundaries].max_boundary = INFINITY;
+
+  return buckets;
+}
+
+distribution_t *distribution_new_linear(size_t num_buckets, double size) {
+  distribution_t *d = (distribution_t *)calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_linear(num_buckets, size);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_exponential(size_t num_buckets,
+                                             double initial_size,
+                                             double factor) {
+  distribution_t *d = (distribution_t *)calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_exponential(num_buckets, initial_size, factor);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_custom(size_t num_boundaries,
+                                        double *custom_buckets_boundaries) {
+  distribution_t *d = (distribution_t *)calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_custom(num_boundaries, custom_buckets_boundaries);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_boundaries;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+static void bucket_update(bucket_t *buckets, size_t num_buckets, double gauge) {
+  int ptr = (int)num_buckets - 1;
+
+  while (buckets[ptr].max_boundary > gauge && ptr >= 0) {
```

Comment:
Unfortunately this for loop doesn't work and will call segmentation fault I think. That's because i is the size_t type and it imply that i will always be greater equal zero. If we subtract 1 from 0, then we will overflow, get the maximum value of size_t and then buckets[i].max_boundary will call seg fault